core/kotlinx-coroutines-debug/src/internal/DebugProbesImpl.kt

```diff
@@ -74,23 +74,30 @@ internal object DebugProbesImpl {
         }
     }
 
+    @Suppress("DEPRECATION_ERROR") // JobSupport
     private fun Job.build(map: Map<Job, CoroutineState>, builder: StringBuilder, indent: String) {
         val state = map[this]
-        builder.append(indent)
-        @Suppress("DEPRECATION_ERROR")
         val str = if (this !is JobSupport) toString() else toDebugString()
-        if (state == null) {
+        val newIndent: String
+        if (state == null) { // Append coroutine without stacktrace
+            // Do not print scoped coroutines and do not increase indentation level
             @Suppress("INVISIBLE_REFERENCE")
-            if (this !is kotlinx.coroutines.internal.ScopeCoroutine<*>) { // Do not print scoped coroutines
-                builder.append("$str\n")
+            if (this !is kotlinx.coroutines.internal.ScopeCoroutine<*>) {
+                builder.append("$indent$str\n")
+                newIndent = indent + "\t"
+            } else {
+                newIndent = indent
             }
         } else {
+            // Append coroutine with its last stacktrace element
             val element = state.lastObservedStackTrace().firstOrNull()
             val contState = state.state
-            builder.append("$str, continuation is $contState at line $element\n")
+            builder.append("$indent$str, continuation is $contState at line $element\n")
+            newIndent = indent + "\t"
         }
+        // Append children with new indent
```

Comment:
The code above is hard to understand to due to tie "print string and increase indent" logic that is duplicated across different if branches. I'd rewrite it somehow. One way (not sure the best one):

```
        val line = if (state == null) { // Append coroutine without stacktrace
            // Do not print scoped coroutines 
            @Suppress("INVISIBLE_REFERENCE")
            if (this is kotlinx.coroutines.internal.ScopeCoroutine<*>) null else "$indent$str\n"
        } else {
            // Append coroutine with its last stacktrace element
            val element = state.lastObservedStackTrace().firstOrNull()
            val contState = state.state
            "$indent$str, continuation is $contState at line $element\n"
        }
        if (line != null) builder.append(line)
        // Append children with new indent
        for (child in children) {
            child.build(map, builder, if (line == null) indent else indent + "\t")
        }
```

Reply:
Still not so readable.
Extracted method

Reply:
Much better